### PR TITLE
Use Microsoft.Extensions.Diagnostics

### LINKED ISSUE #47618: Replace internal Microsoft.Extensions.Metrics with package
https://github.com/dotnet/aspnetcore/pull/46834 adds an internal prototype version of Microsoft.Extensions.Metrics. The types are shared throughout ASP.NET Core with `InternalsVisibleTo`.

Replace with Microsoft.Extensions.Metrics when available. This work is happening in https://github.com/dotnet/runtime/issues/77514

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/47618

@davidfowl @tarekgh @noahfalk

### CODE REVIEW

**Comment by JamesNK** [comment]
A smooth migration (as expected)

Feedback:
* Should `DefaultMeterFactory` be public? It would save having to have a `TestMeterFactory`.
* `InstrumentRecorder<T>` lets you see measurements that have been collected, but there is no notification that a value has been received. There are a couple of tests where I needed to await new values. Does it make sense to have that notification on the recorder.

**Comment by JamesNK** [comment]
> Cannot reference "Microsoft.Extensions.Diagnostics". This dependency is not in the shared framework. See docs/SharedFramework.md for instructions on how to modify what is in the shared framework.

`Microsoft.Extensions.Diagnostics` needs to be in the shared framework so it can be referenced by ASP.NET Core bits in the shared framework (Hosting, Kestrel, etc)

**Comment by JamesNK** [comment]
@wtgodbe This PR adds references to `Microsoft.Extensions.Diagnostics`. Could you double-check the PR and make sure I have performed the necessary rituals?

**Comment by noahfalk** [comment]
> A smooth migration (as expected)

Excellent 👍 

> Should DefaultMeterFactory be public? It would save having to have a TestMeterFactory

serviceProvider.GetRequiredService\<IMeterFactory>() is insufficient?

>  There are a couple of tests where I needed to await new values. Does it make sense to have that notification on the recorder.

Tarek originally had one but I believed it would have no use and suggested we get rid of it. So egg on my face :) Do you have a good example so I can better understand where you found it useful?

**Comment by JamesNK** [comment]
> serviceProvider.GetRequiredService<IMeterFactory>() is insufficient?

That works if you have a DI container setup in your test. Without a DI container, you need to create your own implementation.

> Tarek originally had one but I believed it would have no use and suggested we get rid of it. So egg on my face :) Do you have a good example so I can better understand where you found it useful?

He might have seen it in the ASP.NET Core prototype.

* InstrumentRecorder with a hacky callback API: https://github.com/dotnet/aspnetcore/blob/4e17e96258aebedd51d4dedb8fab201f7f3d600c/src/Shared/Metrics/InstrumentRecorder.cs
* Example of a test that uses it: https://github.com/dotnet/aspnetcore/blob/4e17e96258aebedd51d4dedb8fab201f7f3d600c/src/Middleware/Diagnostics/test/UnitTests/DeveloperExceptionPageMiddlewareTest.cs#L536-L589

I've worked around these two issues (an IMeterFactory implementation for tests and a way to get notifications of new measurements [here](https://github.com/dotnet/aspnetcore/blob/jamesnk/metrics-extensions/src/Shared/Metrics/TestMeterFactory.cs))

File: eng/Dependencies.props [hunk 1]
```diff
@@ -40,6 +40,7 @@ and are generated based on the last package release.
     <LatestPackageReference Include="Microsoft.Extensions.DependencyInjection" />
     <LatestPackageReference Include="Microsoft.Extensions.DependencyModel" />
     <LatestPackageReference Include="Microsoft.Extensions.DiagnosticAdapter" />
+    <LatestPackageReference Include="Microsoft.Extensions.Diagnostics" />
```
Comment (tarekgh) [inline eng/Dependencies.props:43]: Don't you need the Microsoft.Extensions.Diagnostics.Abstraction too? or will the dependency be automatically picked?

Comment (tarekgh) [reply eng/Dependencies.props:43]: CC @joperezr

**Comment by tarekgh** [comment]
> That works if you have a DI container setup in your test. Without a DI container, you need to create your own implementation.

You can see what we are doing in our tests. create a temporary ServiceCollection, call AddMetrics then request IMeterFactory from the service. It doesn't matter if you are running with a container or not.

**Comment by noahfalk** [comment]
> Example of a test that uses it:

```C#
        var response = await server.CreateClient().GetAsync("/path");
        Assert.Equal(HttpStatusCode.InternalServerError, response.StatusCode);
        ...
```

In this example isn't awaiting the GetAsync() call sufficient to know that the work has been done and the measurement is recorded? Is there some asynchronous work happening somewhere even after the await completes?

**Comment by JamesNK** [comment]
> In this example isn't awaiting the GetAsync() call sufficient to know that the work has been done and the measurement is recorded?

In this particular test, there is a race condition. The server can return the failing response to the client before the metric is recorded.

99% of the time it is fine, but 1% of the time the client can report it's finished before the request is finished on the server and the counter is recorded.

**Comment by eerhardt** [comment]
This change was merged into #48407. Closing.

**Comment by JamesNK** [comment]
> > In this example isn't awaiting the GetAsync() call sufficient to know that the work has been done and the measurement is recorded?
> 
> In this particular test, there is a race condition. The server can return the failing response to the client before the metric is recorded.
> 
> 99% of the time it is fine, but 1% of the time the client can report it's finished before the request is finished on the server and the counter is recorded.

Created issue: https://github.com/dotnet/runtime/issues/86783

**Comment by JamesNK** [comment]
> > That works if you have a DI container setup in your test. Without a DI container, you need to create your own implementation.
> 
> You can see what we are doing in our tests. create a temporary ServiceCollection, call AddMetrics then request IMeterFactory from the service. It doesn't matter if you are running with a container or not.

I think this is a little gross. However, it's not a big deal to create a simple `IMeterFactory` if you don't want to involve a container at all. Can leave it as is for now.

**Comment by noahfalk** [comment]
> In this particular test, there is a race condition. The server can return the failing response to the client before the metric is recorded.

AH interesting. In general I wouldn't want to encourage users to write tests with timed waits for a callback if they had an option to avoid it, but I'm fine for us to add a callback API and use docs/comments to provide that guidance.

**Pull request closed without merging**